qiime2/metadata/io.py

```diff
@@ -88,6 +90,10 @@ def read(self, into, column_types=None):
         index = pd.Index(ids, name=header[0], dtype=object)
         df = pd.DataFrame(data, columns=header[1:], index=index, dtype=object)
 
+        # TODO: move these checks over to Metadata.__init__() so that you can
+        # pass column_types with an untyped dataframe. This would require a bit
+        # of a refactor and doesn't buy a whole lot at the moment, hence the
+        # TODO.
```

Comment:
Turns out I need to do this, otherwise the data is cast to numeric before I have a chance to replace to NaN. I could just do the conversion twice, but it would be nice to have this anyhow and it would clean up a few things I think.

Goal will be to retain the current exceptions so I don't have to modify any tests.

Reply:
Will probably try to factor this out into a shared approach rather than shoving it all into __init__